### PR TITLE
Better code.

### PR DESCRIPTION
**Pull request by johanatan**

### CODE REVIEW

**Comment by weavejester** [comment]
Can you explain why this is better code?

**Comment by johanatan** [comment]
It’s 17 vs 22 tokens (excluding punctuation) and captures the duplication in that every special case selects the second item in the tuple. It’s more maintainable (assuming that new special cases will also select the second/new item) and less error prone (as any one of those special cases could accidentally select the first/old value as well).

**Comment by weavejester** [comment]
Avoiding duplication is good, but it's also harder to read. As a rule of thumb I tend to prioritize readability over reducing duplication.

**Comment by johanatan** [comment]
Firstly, deduplication was only one aspect of my answer. The other is that this code actually captures a true relationship that exists.

Secondly, how is this “less readable”? It’s a lookup in a set. That’s readable.

**Comment by weavejester** [comment]
> Firstly, deduplication was only one aspect of my answer.

Sure, but I thought it was your strongest argument.

> Secondly, how is this “less readable”? It’s a lookup in a set. That’s readable.

When I look at the existing code, I see the line:

    (case [old-state new-state]

And I can immediately tell that this is a state machine, with the cases listed below. In other words, I only need to glance over the first three tokens to understand the intent of the entire function.

Now let's take a look at your code:

    (if (#{[nil :pass] [nil :fail] [:pass :fail] [nil :error] [:pass :error] [:fail :error]}

This line tells me nothing on it's own, and we're already at 13 tokens. So onto the next part:

    [old-state new-state])

Okay, so now I can see this is a set lookup on the old and new state, but since it's spread over two lines and the set is so long, it's visually disconnected. It's also less obvious to me that this is a state machine.

Finally:

    new-state old-state))

Now the purpose becomes fully clear, and I think this is the one part that's actually clearer than the existing code. However, it does assume that the state will either be the old or the new state, which may not necessarily be the case in future, so we lose some future proofing.

In summary, while your code is more concise, it also hides the intent of the code. I need to read through more tokens before I understand its purpose.

Readability is of course rather subjective, so you may see things differently. In the case of a disagreement we should optimise toward the person who will be reading and maintaining this code the most... which happens to be me. So unfortunately I have to say that the existing code stands, because for me its more clearly displays the intent.

**Comment by mariusz-jachimowicz-83** [comment]
I like the original `case` like approach. Good code is not always about shorter one. Good code for me:
* easy to eye scan (often with columnar alignment)
* code that don't make/force me to think

**Comment by johanatan** [comment]
Like I mentioned, columnar alignment is possible with my code. But the biggest benefit is that it would prevent accidentally selecting the wrong value for any given case. If such time as a case comes along not confirming, then yea refactor/unroll the loop. Thats just my preference though.
